drivers/flash: deprecation of flash_write_protection_set() API

nvlsianpu opened the pull request:
The API is deprecated as it was decided to build in the flash protection
service into write and erase procedures.
This is solution chosen for fix following issue:
  When two or more threads writes to flash device it is
  possible that flash protection will be enabled by one of
  threads despite another thread(s) still needs it disabled.

fixes #31217 

- [x] spi_flash_at45 (already ready)
- [x] soc_flash_nrf (already ready)
- [x] nrf_qspi_nor
- [x] spi_nor
- [x] soc_flash_lpc
- [x] soc_flash_mcux
- [x] soc_flash_rv32m1
- [x] soc_flash_nios2_qspi
- [x] flash_esp32 (already ready)
- [x] flash_gecko
- [x] flash_mcux_flexspi_nor (already ready)
- [x] flash_sam
- [x] flash_sam0
- [x] flash_stm32_qspi
- [x] flash_stm32
- [x] flash_stm32h7
- [x] flash_simulator
- [x] remove CONFIG_FLASH_SIMULATOR_ERASE_PROTECT (not relevant anymore)
- [x] align tests
- [x] removed usage in tests
- [x] removed usage in samples
- [x] remove usage in drivers, subsystems and wherever else

Additionally:
fs/nvs:
Removed two unneeded `if` statements. These statements had not
effect in function results.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/nrf_qspi_nor: integrate protection service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/spi_nor: integrate protection service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/soc_flash_mcux: integrate protection service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/soc_flash_lpc: integrate protection service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/soc_flash_rv32m1: integrate WP service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/soc_flash_sam: integrate WP service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

Write procedure was fixed so it doesn't lock the driver after
failure anymore.

flash_write_protection_set() will be deprecated.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/soc_flash_sam0: integrate WP service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/flash_gecko: integrate WP service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/flash_stm32: integrate WP service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/flash_stm32h7x: integrate WP service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/flash_stm32_qspi: integrate WP service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/soc_flash_nios2_qspi: integrate WP service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/flash_simulator: integrate WP service into write/erase

Flash write protection services were integrated into erase and write
procedures. This is step required for fixing following issue:
Multi-threading flash access is not supported by
flash_write_protection_set().

flash_write_protection_set() will be deprecated

As CONFIG_FLASH_SIMULATOR_ERASE_PROTECT become a dead option.
this commit removes it as well.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
include/drivers/flash: deprecation of flash_write_protection_set()

The API is deprecated as it was decided to build in the flash protection
service into write and erase procedures.
This is solution chosen for fix following issue:
  When two or more threads writes to flash device it is
  possible that flash protection will be enabled by one of
  threads despite another thread(s) still needs it disabled.

fixes #3127

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash: removed write protection API implementations

flash_write_protection_set() API was deprecated so driver
implementation interface is left behind as well.

This patch removes all implementation pointed by the interface
'struct flash_driver_api.write_protection'.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests/drivers/flash_simulator: remove flash_write_protection_set() usage

flash_write_protection_set() was deprecated.
This patch removes usage of it and the test case which was
testing deprecated API behaviors.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests/tree-wide: remove flash_write_protection_set() usage

flash_write_protection_set() was deprecated.
This patch removes usage of it.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests/boards/altera_max10/qspi: remove flash_write_protection_set() usage

This patch removes scenario which was testing deprecated
API behaviors. Needed as As flash_write_protection_set() was
deprecated and became no-operation.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
release-note: flash_api_write_protection() deprecation record

Added note on deprecation of the API and description of flash
drivers changes.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
fs/nvs: remove flash_write_protect_set() usage

flash_write_protection_set() was deprecated.
This patch removes usage of it and the test case which was
testing deprecated API behaviors.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
fs/nvs: removed unneeded `if` statements

Removed two unneeded `if` statements. These statements had not
effect in function results.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
storage: remove flash_write_protection_set() usage

This patch removes scenario which was testing deprecated
API behaviors. Needed as As flash_write_protection_set() was
deprecated and became no-operation.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
disk/disk_acces_flash: remove flash_write_protection_set() usage

This patch removes scenario which was testing deprecated
API behaviors. Needed as As flash_write_protection_set() was
deprecated and became no-operation.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/flash_shell: remove flash_write_protection_set() usage

This patch removes scenario which was testing deprecated
API behaviors. Needed as As flash_write_protection_set() was
deprecated and became no-operation.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/eeprom/eeprom_emulator: remove flash_write_protection_set() usage

This patch removes scenario which was testing deprecated
API behaviors. Needed as As flash_write_protection_set() was
deprecated and became no-operation.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
samples/tree-wide: remove flash_write_protection_set() usage

This patch removes scenario which was testing deprecated
API behaviors. Needed as As flash_write_protection_set() was
deprecated and became no-operation.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -419,8 +417,6 @@ static int flash_stm32h7_write_protection(const struct device *dev, bool enable)
 		LOG_DBG("Disable write protection");
 	}
 
-	flash_stm32_sem_give(dev);
```
yeah, good catch

pabigot commented on include/drivers/flash.h:
```diff
@@ -254,16 +251,17 @@ static inline int z_impl_flash_erase(const struct device *dev, off_t offset,
+	ARG_UNUSED(enable);
 
-	return api->write_protection(dev, enable);
+	return 0;
```
It strikes me that making this function a no-op will break out-of-tree flash drivers that do not yet incorporate automatic write protection management, which isn't quite the intent behind deprecation.

nvlsianpu replied:
I'm not sure whether zephyr should thread drivers shim implementation API similar as driver user API.
For my point of view `api->write_protection` is something internal (so I don't care out-off-tree driver).
@carlescufi Any thought on that?

nvlsianpu replied:
Will ask @anangl I bet he has an experience on such things.

pabigot commented on drivers/flash/spi_nor.c:
```diff
@@ -480,36 +483,43 @@ static int spi_nor_write(const struct device *dev, off_t addr,
 
 	acquire_device(dev);
+	ret = spi_nor_write_protection_set(dev, false);
+	if (ret == 0) {
```
I think it would be easier to review if this didn't shift the entire 30+ line while loop over in order to stick in the conditional clause.  A `goto out;` if write protection setting failed seems reasonable.

nvlsianpu replied:
Personally I prefer to avoid `goto`. I did similar changes In a few places more. Do you want that I apply your comments to them all?

pabigot replied:
You don't have to do it anywhere, it's just making things harder to review.  IMO.

de-nordic replied:
The `ret == 0` could just be taken into the `while` condition, this would reduce the change to single line, no need to indent code.

pabigot commented on doc/releases/release-notes-2.6.rst:
```diff
@@ -148,6 +150,12 @@ Drivers and Sensors
 * Flash
 
+  * flash_write_protection_set() was deprecated API and become no-operation.
+    For all the driver service of flash write protectin was embedded into write
```
s/protectin/protection/

pabigot commented on include/drivers/flash.h:
```diff
@@ -166,9 +163,6 @@ static inline int z_impl_flash_read(const struct device *dev, off_t offset,
  *  Write size and offset must be multiples of the minimum write block size
  *  supported by the driver.
  *
- *  Prior to the invocation of this API, the flash_write_protection_set needs
```
This shouldn't just be dropped; it should be replaced with text that specifies that any necessary write protection management must be performed by the driver, with the driver responsible for ensuring the "write-protect" after the operation completes (successfully or not) matches the write-protect state when the operation was started.

Otherwise there isn't enough detail here to tell an implementer what this has to do.

pabigot commented on include/drivers/flash.h:
```diff
@@ -224,6 +215,9 @@ static inline int z_impl_flash_erase(const struct device *dev, off_t offset,
  *  @brief  Enable or disable write protection for a flash memory
  *
+ *  This API is deprecated and will be removed in Zephyr 2.8.
+ *  It will be keep as No-Operation until that.
```
This too should mention that the responsibility for this function is now transferred to the write and erase operations.

pabigot reviewed (dismissed):
A few high-level comments.  I think this is a reasonable approach, but I'm not sure it can be done without breaking working systems (especially out-of-tree drivers).

Blocking comments are related to the documentation of the API; the rest is commentary.

nvlsianpu committed:
include/drivers/flash: make write_protection handler optional

Made write_protection handler not mandatory.
flash_write_protection_set() becomes no-operation.
If write_protection handler is provided by the driver implementation
it will be called within flash_writ() and flash_erase().

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

chrta commented on doc/releases/release-notes-2.6.rst:
```diff
@@ -159,6 +161,12 @@ Drivers and Sensors
+  * flash_write_protection_set() was deprecated API and become no-operation.
+    For all the driver service of flash write protection was embedded into write
+    and erase operations implementations. Driver implementation interface API
+    :c:type:`flash_api_write_protection` becomes unneeded anymore, and was
```
I think instead of `becomes unneeded anymore`, `is not needed anymore` or `becomes unneeded` would be better.

nvlsianpu replied:
Actually text was incorrect.

chrta commented on drivers/flash/flash_gecko.c:
```diff
@@ -109,20 +112,27 @@ static int flash_gecko_erase(const struct device *dev, off_t offset,
 }
 
-static int flash_gecko_write_protection(const struct device *dev, bool enable)
+static int flash_gecko_write_protection_nop(const struct device *dev,
```
This function seems to be unused

nvlsianpu commented:
@pabigot @chrta @de-nordic Please revisit in your convenience.

chrta commented on drivers/flash/soc_flash_lpc.c:
```diff
@@ -61,7 +61,7 @@ static int flash_lpc_erase(const struct device *dev, off_t offset, size_t len)
 	uint32_t page_size;
 
-	if (k_sem_take(&priv->write_lock, K_NO_WAIT)) {
+	if (k_sem_take(&priv->write_lock,  K_FOREVER)) {
```
```suggestion
	if (k_sem_take(&priv->write_lock, K_FOREVER)) {
```

nvlsianpu replied:
fixed

chrta reviewed (changes requested):
Just one minor nit. Otherwise looks good to me. I successfully tested this doing a brief test with samples/drivers/flash_shell on the gecko board efr32mg_sltb004a

Laczen reviewed (changes requested):
Regarding out of tree flash drivers, as they might still rely on `flash_api_write_protection` all changes to the use of flash driver: eeprom_emulator, nvs, shell, disk, flash_map, stream_flash, tests, fcb, ... should be removed from the PR.

nvlsianpu commented:
@Laczen Deprecation process guide demands:
> Code using the deprecated API needs to be modified to remove usage of said API

https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#deprecated

Laczen commented:
> @Laczen Deprecation process guide demands:
> 
> > Code using the deprecated API needs to be modified to remove usage of said API
> 
> https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#deprecated

Something is not right here: one the one hand we say that we keep `flash_write_protection_set` for compatibility with out of tree drivers, while on the other hand it is removed from usage. Then there will be out of tree drivers that are compatible, but they might not work.
Will there be a warning when a project is build with an out of tree driver that has the `flash_write_protection_set` defined but is using e.g. nvs that is no longer calling `flash_write_protection_set` ?

nvlsianpu commented:
I made the note which said:
> Driver implementation interface API :c:type:`flash_api_write_protection` becomes unneeded internally, although is kept for ensuring out-of-tree driver code compatibility.

Code compatible mean that these drivers will still be able to be build.

> Will there be a warning when a project is build with an out of tree driver that has the flash_write_protection_set defined but is using e.g. nvs that is no longer calling flash_write_protection_set?

No, although would be great if can generate a (build time?) warning if a driver provide a `write_protect` handler, but that not easy.

nvlsianpu commented:
^^ @Laczen will propose additional patch

nvlsianpu commented:
@Laczen  What about patch to write and erase API implementations?
``` C
static inline int z_impl_flash_write(const struct device *dev, off_t offset,
				     const void *data, size_t len)
{
	const struct flash_driver_api *api =
		(const struct flash_driver_api *)dev->api;
	int rc;

	if (api->write_protection != NULL) {
		rc = api->write_protection(dev, false);
		if (rc) {
			return rc;
		}
	}

	rc = api->write(dev, offset, data, len);

	if (api->write_protection != NULL) {
		(void)api->write_protection(dev, true);
	}

	return rc;
}
```

chrta reviewed (commented):
lgtm

Laczen reviewed (approved):
Yes, this should work. LGTM

de-nordic commented on drivers/flash/flash_sam0.c:
```diff
@@ -238,23 +240,34 @@ static int flash_sam0_write(const struct device *dev, off_t offset,
+		flash_sam0_commit(dev);
+	}
+
+	int err2 = flash_sam0_write_protection(dev, true);
```
Shouldn't this be also inside the `if (err == 0)` block? If `err != 0` we have failed to unlock the flash anyway, the lock will probably also fail.

nvlsianpu replied:
My point is to not leave memory as unprotected if possible. It might be (quite probably) that it not a case after unprotect call failed, but calling the protect shouldn't make any additional harm.

de-nordic reviewed (approved):
Looks OK.

Left some notes.
I also have some doubt whether applying lock when we have failed to unlock the flash at first, in error processing paths, make sense;  it would probably result in additional error anyway, but since we are failing, maybe additional `if` just to avoid it does not make sense.

nvlsianpu commented:
@de-nordic see my comment. I think call protect doesn't make any harm.

de-nordic commented:
> 
> 
> @de-nordic see my comment. I think call protect doesn't make any harm.

Yes. I have seen it. I have already approved: I myself am not sure whether having branch over the lock makes sense if we are already failing, so no harm if we just attempt lock and fail again.

nvlsianpu commented:
@pabigot Can you revisit.

pabigot commented on doc/releases/release-notes-2.6.rst:
```diff
@@ -184,6 +186,12 @@ Drivers and Sensors
 
 * Flash
 
+  * flash_write_protection_set() was deprecated API and become no-operation.
```
While this isn't wrong, it confused me because it only explains part of what has been and what needs to be done. 
 People reading this may not know what they need to do to update out-of-tree drivers and their code.

I drafted this which is closer, but probably not quite right.  It should be reconciled with what's in the other comment in this review.

> `flash_write_protection_set()` has been deprecated and will be removed in Zephyr 2.8.  Responsibility for write/erase protection management has been moved to the driver-specific implementation of the `flash_write()` and `flash_erase()` API calls.  All in-tree flash drivers have been updated, and the protect implementation removed from their API tables. During the deprecation period user code invoking `flash_write_protection_set()` will have no effect, but the `flash_write()` and `flash_erase()` driver shims will wrap their calls with calls to the protect implementation if it is present in the API table.  Out-of-tree drivers must be updated before the wrapping in the shims is removed when the deprecation period ends.

nvlsianpu replied:
Applied

pabigot reviewed (commented):
AFAICT this probably satisfies my concerns about backwards-compatible behavior, but I find the description of the changes to be a little confusing.  Please review the suggestions; this can go in as-is, or if changes are made that remove my confusion I'll approve.

nvlsianpu commented:
thanks @pabigot I will take into accounts your remarks.

nvlsianpu commented:
@pabigot ?

pabigot commented on doc/releases/release-notes-2.6.rst:
```diff
@@ -184,6 +186,18 @@ Drivers and Sensors
+    During the deprecation period user code invoking
+    flash_write_protection_set() will have no effect, but the flash_write() and
+    flash_erase() driver shims will wrap their calls with calls to the protect
+    implementation if it is present in the API table
```
Missing `.` at end of sentence.

nvlsianpu commented:
@nashif @carlescufi @MaureenHelm

nvlsianpu committed:
tests/drivers/flash: allow CI using nrf52840dk_nrf52840

Added integration_platforms with nrf52840dk_nrf52840.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
DNM, I found nrf_qspi_nor issue

nvlsianpu commented:
Fixed nrf_qspi_nor issue by intorduced additional locking mechanism for `unprotect->write->protect` & `unprotect->erase->protect` transaction. E.g. write and protect services have been using same resource lock before which has made operation impossible.

nvlsianpu committed:
drivers/flash/flash_stm32g4x: fixed build issue

This driver file couldn't be compiled due the syntax error.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
